subsys/bluetooth/host/l2cap.c

```diff
@@ -1256,6 +1252,17 @@ static void l2cap_chan_seg_sent(struct bt_conn *conn, void *user_data)
 	l2cap_chan_tx_resume(BT_L2CAP_LE_CHAN(chan));
 }
 
+static bool test_and_dec(atomic_t *target)
+{
+	if (!atomic_get(target)) {
+		return false;
+	}
+
```

Comment:
I think we have a race condition here in between the check to see if target is nonzero, and decrementing the target, you could get preempted in between, or it gets updated on another CPU at the same time.

I recommend you look at lib/os/sem.c, note the use of atomic_cas() with while loops.

Reply:
Interesting if such construction is recommended why we don't we promote them to atomic_bound_dec and atomic_bound_inc? They could be used as is in this case.

Reply:
(Oops, fat-fingered the resolve button, sorry).

To be clear: that construction is required.  Consider the case where you enter this function, decide that the count is non-zero, and then between the if() block and the atomic_dec() another thread runs and decrements the count to zero.  This line will then wrap it around to 0xffffffff incorrectly.   Two atomic operations used sequentially is not atomic.

I don't think having a bounded add/dec operation in the kernel would be a bad idea at all, though it's a little out of scope of the existing atomic.h API which is really there just to expose common/portable hardware instructions for atomic operations.

Reply:
the bounded_inc/bounded_dec functions have some special semantics related to how it is used with futexes (bounded dec will decrement to minimum -1 and not minimum).

So I'd be very careful in making sure that this works precisely the way you want if you use that code unmodified.

However, looping with atomic_cas() is a general design pattern to make sure that the value wasn't changed underneath you while you were working on it, so whatever you come up with I think would surely look similar.

Reply:
Yep, Ive added a pretty similar loop, it just doesn't take a minimum value since all we care is that it doesn't decrement bellow zero.